#### 图片信息获取（7.15+太原理工荟萃文创+132）

In [3]:
import cv2
import matplotlib.pyplot as plt
import matplotlib.image as mp
import numpy as np
import os
import csv
from PIL import Image
from skimage.measure import compare_ssim
import argparse
import imutils

In [20]:
# 获取文件中的图片数

# 文件夹路径
path ="./chushi"
# 打开对应的文件夹
filelist = os.listdir(path)  # 使用os库
# 得到文件数
num = len(filelist)

In [18]:
# 把图片信息存入csv

# 1. 创建文件对象
f = open('photo_info2.csv','a',newline='',encoding='utf-8')

# 2. 基于文件对象构建 csv写入对象
csv_writer = csv.writer(f)

# 3. 构建列表头
csv_writer.writerow(["图片名","图像的格式","图像的大小","图像的宽度","图像的高度","获取某个像素点的颜色值"])


41

In [21]:
# 获取图片信息并存入上述文件夹

# 遍历文件数的次数
for i in range(num):
    # 打开图片
    im2 = Image.open('./chushi/mmexport ('+str(i+1)+').jpg')
    # 显示图片
    # im2.show()
    # 导出信息到csv
    csv_writer.writerow(['mmexport('+str(i+1)+')',im2.format,im2.size,im2.width,im2.height,im2.getpixel((100, 100))])
    # 验证信息
    print('图像的格式：', im2.format)
    print('图像的大小：', im2.size)
    print('图像的宽度：', im2.width)
    print('图像的高度：', im2.height)
    # 传入坐标的元组
    print('获取某个像素点的颜色值：', im2.getpixel((100, 100)))

图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 2340)
图像的宽度： 1080
图像的高度： 2340
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (234, 234, 234)
图像的格式： JPEG
图像的大小： (360, 800)
图像的宽度： 360
图像的高度： 800
获取某个像素点的颜色值： (237, 241, 244)
图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： PNG
图像的大小： (1242, 2208)
图像的宽度： 1242
图像的高度： 2208
获取某个像素点的颜色值： (238, 238, 238)
图像的格式： PNG
图像的大小： (1125, 2001)
图像的宽度： 1125
图像的高度： 2001
获取某个像素点的颜色值： (235, 235, 235, 255)
图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (238, 238, 238)
图像的格式： JPEG
图像的大小： (1080, 2400)
图像的宽度： 1080
图像的高度： 2400
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 2340)
图像的宽度： 1080
图像的高度： 2340
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 2340)
图像的宽度： 1080
图像的高度： 2340
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (526, 1125)
图像的宽度： 526
图像的高度： 1125
获取某个像素点的颜色值：

图像的宽度： 1080
图像的高度： 2340
获取某个像素点的颜色值： (18, 18, 18)
图像的格式： JPEG
图像的大小： (720, 1436)
图像的宽度： 720
图像的高度： 1436
获取某个像素点的颜色值： (96, 96, 96)
图像的格式： JPEG
图像的大小： (828, 1792)
图像的宽度： 828
图像的高度： 1792
获取某个像素点的颜色值： (235, 235, 235)
图像的格式： JPEG
图像的大小： (720, 1280)
图像的宽度： 720
图像的高度： 1280
获取某个像素点的颜色值： (248, 248, 248)
图像的格式： JPEG
图像的大小： (1080, 2339)
图像的宽度： 1080
图像的高度： 2339
获取某个像素点的颜色值： (235, 235, 235)
图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (238, 238, 238)
图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 2312)
图像的宽度： 1080
图像的高度： 2312
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (720, 1520)
图像的宽度： 720
图像的高度： 1520
获取某个像素点的颜色值： (239, 239, 239)
图像的格式： JPEG
图像的大小： (1080, 1920)
图像的宽度： 1080
图像的高度： 1920
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1176, 2400)
图像的宽度： 1176
图像的高度： 2400
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 2340)
图像的宽度： 1080
图像的高度： 2340
获取某个像素点的颜色值： (237, 237, 237)
图像的格式： JPEG
图像的大小： (1080, 

结果： 得到132张图片信息（图像的格式、大小、宽度、高度、获取某个像素点的颜色值）

#### 重合度

In [34]:
# 把重合度信息输入到列表
list = []

In [35]:
# 把每张照片和其他照片对比查重

# 总遍历132次
for i in range(num):
    try:
        # 获取图片名
        jpg_name = path + '/mmexport ('+str(i+1)+').jpg'
        # 导入图片
        image = mp.imread(jpg_name)
        # 将图片大小统一
        image_size = cv2.resize(image, (1440 , 2400))
        # 将图片灰度处理
        image_gray = cv2.cvtColor(image_size,cv2.COLOR_BGR2GRAY)
        # 验证图片名
        print(jpg_name)
        
        # 每张图片都与其他131张图片进行遍历对比
        for ii in range(num):
            # 同上代码
            jpg_name_comp = path + '/mmexport ('+str(ii+1)+').jpg'
            # 筛选出非自身的图片
            if(jpg_name_comp !=jpg_name):
                image_comp = mp.imread(jpg_name_comp)
                image_size_comp = cv2.resize(image_comp, (1440 , 2400))
                image_gray_comp = cv2.cvtColor(image_size_comp,cv2.COLOR_BGR2GRAY)
                
                # 通过compare_ssim方法对比重合度
                (score , diff) = compare_ssim(image_gray,image_gray_comp,full = True)
                # 转化中文
                diff = (diff * 255).astype("uint8")
                # 把信息放入list
                list.append(score)
    except:
        pass

./chushi/mmexport (1).jpg


F:\anaconda\lib\site-packages\ipykernel_launcher.py:14: UserWarning: DEPRECATED: skimage.measure.compare_ssim has been moved to skimage.metrics.structural_similarity. It will be removed from skimage.measure in version 0.18.
  


./chushi/mmexport (2).jpg
./chushi/mmexport (3).jpg
./chushi/mmexport (4).jpg
./chushi/mmexport (5).jpg
./chushi/mmexport (6).jpg
./chushi/mmexport (7).jpg
./chushi/mmexport (8).jpg
./chushi/mmexport (9).jpg
./chushi/mmexport (10).jpg
./chushi/mmexport (11).jpg
./chushi/mmexport (12).jpg
./chushi/mmexport (13).jpg
./chushi/mmexport (14).jpg
./chushi/mmexport (15).jpg
./chushi/mmexport (16).jpg
./chushi/mmexport (17).jpg
./chushi/mmexport (18).jpg
./chushi/mmexport (19).jpg
./chushi/mmexport (20).jpg
./chushi/mmexport (21).jpg
./chushi/mmexport (22).jpg
./chushi/mmexport (23).jpg
./chushi/mmexport (24).jpg
./chushi/mmexport (25).jpg
./chushi/mmexport (26).jpg
./chushi/mmexport (27).jpg
./chushi/mmexport (28).jpg
./chushi/mmexport (29).jpg
./chushi/mmexport (30).jpg
./chushi/mmexport (31).jpg
./chushi/mmexport (32).jpg
./chushi/mmexport (33).jpg
./chushi/mmexport (34).jpg
./chushi/mmexport (35).jpg
./chushi/mmexport (36).jpg
./chushi/mmexport (37).jpg
./chushi/mmexport (38).jpg
./chushi/

In [36]:
# 导出文本

# 转化字符串
list_str = str(list)
# 新建txt文本
f=open("comp.txt","w")
# 放入信息
f.writelines(list_str)
# 关闭文本
f.close()

In [77]:
# 显示查重信息

# 记录总重复数
all = 0
# 遍历132 * 131次
for i in range(17292):
    try:
        # 判断重合度大于95%为高重合度
        if(list[i]>0.95):
            # 获取首图
            first = int(i/131)
            # 获取对比图
            sec = i-131*int(i/131)
            # 对比图位置一定大于首图
            if(sec >= first):
                # 记录+1
                all +=1
                # 输出信息
                print('重合度95%以上：图片',first,'图片',sec+1,'  重合度',list[i])
    except:
        pass
print('总：',all)

重合度95%以上：图片 13 图片 20   重合度 0.9705493760585925
重合度95%以上：图片 13 图片 28   重合度 0.9699750709806011
重合度95%以上：图片 19 图片 72   重合度 0.9791913473165149
重合度95%以上：图片 19 图片 73   重合度 0.9656156692506226
重合度95%以上：图片 20 图片 28   重合度 0.9592562169402861
重合度95%以上：图片 23 图片 25   重合度 0.9840649985021684
重合度95%以上：图片 23 图片 35   重合度 0.9818068208363775
重合度95%以上：图片 23 图片 75   重合度 0.9686684818483222
重合度95%以上：图片 23 图片 87   重合度 0.9639277648434548
重合度95%以上：图片 25 图片 35   重合度 0.9846983077367418
重合度95%以上：图片 25 图片 75   重合度 0.971531898695815
重合度95%以上：图片 25 图片 87   重合度 0.9635566240467628
重合度95%以上：图片 29 图片 32   重合度 0.967603463136202
重合度95%以上：图片 31 图片 119   重合度 0.985268365193711
重合度95%以上：图片 34 图片 96   重合度 0.9571747774595059
重合度95%以上：图片 35 图片 75   重合度 0.9705859041984134
重合度95%以上：图片 35 图片 87   重合度 0.9637314458278303
重合度95%以上：图片 46 图片 81   重合度 0.9759353565838146
重合度95%以上：图片 46 图片 104   重合度 0.9751850732270018
重合度95%以上：图片 51 图片 60   重合度 0.9663326748762142
重合度95%以上：图片 52 图片 59   重合度 0.9767151901706853
重合度95%以上：图片 55 图片 63   重合度 0.971206

In [3]:
#加载原始RGB图像
img_rgb = cv2.imread("./ori/mmexport1594864858440.jpg")
#创建一个原始图像的灰度版本，所有操作在灰度版本中处理，然后在RGB图像中使用相同坐标还原
img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_BGR2GRAY)

结果： 得到132 * 131个重合度的值，并导出txt文本，通过分析有29张图片（22%）重合度高于95%，可判断该文件中有78%图片无效。